Mathlib/Tactic/ExtractGoal.lean

```diff
@@ -0,0 +1,128 @@
+/-
+Copyright (c) 2017 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Simon Hudon, Kyle Miller, Damiano Testa
+-/
+import Std.Lean.Meta.Inaccessible
+
+/-!
+#  `extract_goal`: Format the current goal as a stand-alone example
+
+Useful for testing tactics or creating
+[minimal working examples](https://leanprover-community.github.io/mwe.html).
+
+```lean
+example (i j k : Nat) (h₀ : i ≤ j) (h₁ : j ≤ k) : i ≤ k := by
+  extract_goal
+
+/-
+theorem extracted_1 (i j k : Nat) (h₀ : i ≤ j) (h₁ : j ≤ k) : i ≤ k := sorry
+-/
+```
+
+* TODO: Add tactic code actions?
+* Output may produce lines with more than 100 characters
+
+### Caveat
+
+The extracted goal may depend on imports, since it relies on delaboration(?).
+Also, the extracted goal may not be equivalent to the given goal.
+
+```lean
+-- `theorem int_eq_nat` is the output of the `extract_goal` from the example below
+-- the type ascription is removed and the `↑` is replaced by `Int.ofNat`:
+-- Lean infers the correct (false) statement
+theorem int_eq_nat {z : Int} : ∃ n, Int.ofNat n = z := sorry
+
+example {z : Int} : ∃ n : Nat, ↑n = z := by
+  extract_goal  -- produces `int_eq_nat`
+  apply int_eq_nat  -- works
+```
+
+However, importing `Std.Classes.Cast`, makes `extract_goal` produce a different theorem
+
+```lean
+import Std.Classes.Cast
+
+-- `theorem extracted_1` is the output of the `extract_goal` from the example below
+-- the type ascription is erased and the `↑` is untouched:
+-- Lean infers a different statement, since it fills in `↑` with `id` and uses `n : Int`
+theorem extracted_1 {z : Int} : ∃ n, ↑n = z := ⟨_, rfl⟩
+
+example {z : Int} : ∃ n : Nat, ↑n = z := by
+  extract_goal
+  apply extracted_1
+/-
+tactic 'apply' failed, failed to unify
+  ∃ n, n = ?z
+with
+  ∃ n, ↑n = z
+z: Int
+⊢ ∃ n, ↑n = z
+-/
+```
+
+Similarly, the extracted goal may fail to type-check:
+```lean
+example (a : α) : ∃ f : α → α, f a = a := by
+  extract_goal
+  exact ⟨id, rfl⟩
+
+theorem extracted_1.{u_1} {α : Sort u_1} (a : α) : ∃ f, f a = a := sorry
+-- `f` is uninterpreted: `⊢ ∃ f, sorryAx α true = a`
+```
+and also
+```lean
+import Mathlib.Data.Polynomial.Basic
+
+--  The `extract_goal` below produces this statement:
+theorem extracted_1 : X = X := sorry
+-- Yet, Lean is unable to figure out what is the coefficients Semiring for `X`
+/-
+typeclass instance problem is stuck, it is often due to metavariables
+  Semiring ?m.28495
+-/
+
+example : (X : Nat[X]) = X := by
+  extract_goal
+```
+-/
+
+open Lean Elab Tactic
+
+/--
+`extract_goal` formats the current goal as a stand-alone theorem or definition.
+It tries to produce an output that can be copy-pasted and just work.
```

Comment:
May as well put a warning that people shouldn't be *too* hopeful :wink: :
```suggestion
It tries to produce an output that can be copy-pasted and just work,
but its success depends on whether the expressions are amenable
to being unambiguously pretty printed.
```

I haven't tested it, but could you try it and if it works add this suggestion?
```
The tactic responds to pretty printing options.
For example, `set_option pp.all in extract_goal` gives the `pp.all` form.
```

Maybe we should add to the first line of the docstring too:
```
`extract_goal` formats the current goal as a stand-alone theorem or definition,
and `extract_goal name` uses the name `name` instead of an autogenerated one.
```

Reply:
The `set_option pp.all in extract_goal` trick is great!!

This works with no errors:
```lean
theorem extracted_1.{u_1} {α : Sort u_1} (a : α) : @Exists.{u_1} (α → α) fun (f : α → α) ↦ @Eq.{u_1} α (f a) a := sorry

example (a : α) : ∃ f : α → α, f a = a := by
  set_option pp.all true in
  extract_goal
  apply extracted_1

open Polynomial

theorem extracted_2 :
  @Eq.{1} (@Polynomial.{0} Nat Nat.semiring) (@Polynomial.X.{0} Nat Nat.semiring)
    (@Polynomial.X.{0} Nat Nat.semiring) := sorry


example : (X : Nat[X]) = X := by
  set_option pp.all true in
  extract_goal
  apply extracted_2
```

Reply:
I added the `pp.all` trick to the docs and a test.

Reply:
Btw, I checked and the polynomial example, failed also in Lean3.  However, the "exists" example actually inserted the type-ascription:
```lean
example {z : ℤ}: ∃ n : ℕ, z = n := begin
  extract_goal coes
end

-- produces:
lemma coes {z : ℤ} : ∃ (n : ℕ), z = ↑n :=
begin
  admit,
end
```
In either case, you could place `set_option pp.all true` before the *statement* and it would print it with that option.  It does not seem to be possible to use `pp.all` *inside* the proof, though.

Reply:
Exists doesn't show a type ascription in Lean 4 because it's implemented using an `app_unexpander` and is transforming a `fun`, but by default lambdas don't show types. I've been wanting to get around to fixing this for a while.